### PR TITLE
Fix LayerShellV1 flicker

### LINKED ISSUE #3450: `wbg` constantly resizing after switching to 2.17.0
I'm using [`wbg`](https://codeberg.org/dnkl/wbg) as my wallpaper app. On Mir 2.15.0, this worked fine in Miriway:
```
shell-component=wbg /path/to/wallpaper.png
```

After switching to 2.17.0, the wallpaper is now constantly flickering between two sizes, and `wbg` logs:
```
machine # [   16.456052] xsession[1033]: info: main.c:119: render: 640x480 (scaled from 1920x1080)
machine # [   16.504047] xsession[1033]: info: main.c:119: render: 1280x800 (scaled from 1920x1080)
machine # [   16.521671] xsession[1033]: info: main.c:119: render: 640x480 (scaled from 1920x1080)
machine # [   16.570346] xsession[1033]: info: main.c:119: render: 1280x800 (scaled from 1920x1080)
```

<details>
  <summary>Bisect log</summary>

```
git bisect start
# Status: warte auf guten und schlechten Commit
# good: [65a0f9c1a86130a000d0845de86bb65e61f6ed10] Merge pull request #3009 from MirServer/fix-alpine-edge
git bisect good 65a0f9c1a86130a000d0845de86bb65e61f6ed10
# Status: warte auf schlechten Commit, 1 guter Commit bekannt
# bad: [a3c667ae3339967febda1a5c97659c313c7639de] ci: devel != noble any more
git bisect bad a3c667ae3339967febda1a5c97659c313c7639de
# bad: [0704026bd06372ea8286a46d8c939286dd8a8c68] Merge pull request #3138 from MirServer/fix-miral-ExternalClient-tests
git bisect bad 0704026bd06372ea8286a46d8c939286dd8a8c68
# skip: [525bf195cb6f82e52eb38ac601de967b04d7d0de] Apply refactorings that happened on target branch
git bisect skip 525bf195cb6f82e52eb38ac601de967b04d7d0de
# skip: [a1de2788676959699d1ed962ead7c6b40a76ddae] Revert "tests: add eglinfo dependency"
git bisect skip a1de2788676959699d1ed962ead7c6b40a76ddae
# skip: [399f38c8ff52fa2120a99134814dc14bb928b8ed] Fix nitpicking
git bisect skip 399f38c8ff52fa2120a99134814dc14bb928b8ed
# good: [af63ec11aaca3c05ba93f10e28378051b6fb2746] WIP but at least we're not crashing
git bisect good af63ec11aaca3c05ba93f10e28378051b6fb2746
# skip: [3ef63bb4ebfa1ccce03484aa48a0a6b75df40b53] Catching VT-switch errors
git bisect skip 3ef63bb4ebfa1ccce03484aa48a0a6b75df40b53
# bad: [389033c113877fa0c8cca61161193ff40d9fa23f] Merge pull request #3104 from MirServer/fix-scaled-outputs
git bisect bad 389033c113877fa0c8cca61161193ff40d9fa23f
# skip: [b3e4da9a10513452c85dea6033d84fe6861e60bd] Resetting an accidentally changed file
git bisect skip b3e4da9a10513452c85dea6033d84fe6861e60bd
# skip: [26873dcc0a83b6f29aa75bd946d77f66a74f63c9] Changing the platform name to virtual instead of virt
git bisect skip 26873dcc0a83b6f29aa75bd946d77f66a74f63c9
# good: [06f516cb6b8ef268b228a8a8ddf6be6925afa31f] Building implementation of the CRTC-only smooth boot support
git bisect good 06f516cb6b8ef268b228a8a8ddf6be6925afa31f
# skip: [3d073f6419c5c94822ca7f3b5d15f82f2bda655b] Merge remote-tracking branch 'origin/main' into platform-API-merge
git bisect skip 3d073f6419c5c94822ca7f3b5d15f82f2bda655b
# good: [989ad916c4f8448935bed16a8c5695063789251e] sanitizers: use modern CMake commands
git bisect good 989ad916c4f8448935bed16a8c5695063789251e
# skip: [12ea68ab6a2a2636b412f0ebec0dcc20f279fbfe] linux_dmabuf: use eglDestroyImageKHR extension (#3069)
git bisect skip 12ea68ab6a2a2636b412f0ebec0dcc20f279fbfe
# skip: [b09b8cadc34e9d413298d125789a0a40bcd4b7c0] Proper const styles for virtual display.cpp
git bisect skip b09b8cadc34e9d413298d125789a0a40bcd4b7c0
# bad: [cc2b82d4df57e29d05f4ba70fdc46c3e6edb7c46] Initial draft of the architecture diagram
git bisect bad cc2b82d4df57e29d05f4ba70fdc46c3e6edb7c46
# bad: [a1f85c3f5689d21693f95193ea21b7b95dc6b5b5] Update obsolete instructions
git bisect bad a1f85c3f5689d21693f95193ea21b7b95dc6b5b5
# good: [7d67a3378cd22f5d8cf2b3442f9d64999d231bfe] Merge pull request #3030 from MirServer/fix-2011
git bisect good 7d67a3378cd22f5d8cf2b3442f9d64999d231bfe
# good: [d0407f7c0ada76e1c03929799dde53977fa562a1] Fix scaling of wayland platform pointer
git bisect good d0407f7c0ada76e1c03929799dde53977fa562a1
# good: [156cc32e95dcbd2a598685e5797da875df449075] Merge pull request #3039 from MirServer/build-snap-release
git bisect good 156cc32e95dcbd2a598685e5797da875df449075
# bad: [6b2aa7846fa4569873c6413585604dba16ad12cd] Fix resize handling
git bisect bad 6b2aa7846fa4569873c6413585604dba16ad12cd
# first bad commit: [6b2aa7846fa4569873c6413585604dba16ad12cd] Fix resize handling
```

</details>

Behaving like this since 6b2aa7846fa4569873c6413585604dba16ad12cd, behaviour is the same on current main (0980da33355a85e45ace121e21bce9ccbad8a64a).

Here's a log of `env WAYLAND_DEBUG=1 wbg /path/to/wallpaper`, in case the extra insight helps: [wbg.log](https://github.com/user-attachments/files/16019677/wbg.log)

### PR DESCRIPTION
**Pull request by tarek-y-ismail**
closes #3450 

We now ignore all LayerShellV1 configure requests except the most recent one.

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
There are failing tests:

```
2024-07-23T14:57:49.4182258Z [  FAILED  ] LayerSurfaceTest.gets_configured_with_supplied_size_even_when_anchored_to_edges
...
2024-07-23T14:57:49.4369420Z [  FAILED  ] Anchor/LayerSurfaceLayoutTest.is_initially_positioned_correctly_for_anchor/3, where GetParam() = Anchor{bottom}, Margin{l: 6, r: 9, t: 12, b: 15} (30 ms)
...
```

**Comment by tarek-y-ismail** [comment]
Seems like the second "fix" breaks more tests than it fixes...

**Comment by AlanGriffiths** [comment]
As discussed (in the bug) I don't think this addresses the problem

File: src/server/frontend_wayland/layer_shell_v1.cpp [hunk 1]
```diff
@@ -585,19 +585,20 @@ void mf::LayerSurfaceV1::get_popup(struct wl_resource* popup)
...
-        inflight_configures.pop_front();
+        return;
     }
 
-    if (inflight_configures.empty() || inflight_configures.front().first != serial)
+    if (!inflight_configure || inflight_configure->first != serial)
     {
         BOOST_THROW_EXCEPTION(std::runtime_error(
             "Could not find acked configure with serial " + std::to_string(serial)));
     }
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/layer_shell_v1.cpp:598]: OK, so you asked why there was a queue, and this error answers: we want to error if the serial doesn't match a pending configure.

But I think the existing logic is wrong anyway:

1.It assumes that serial never overflows; and,
2. It processes stale acks (that you do fix)

So I think that

1. The queue should stay; and,
2. The logic should be something like...

```c++
    auto const acked_event = std::find_if(
        std::begin(inflight_configures),
        std::end(inflight_configures),
        [serial](auto& p) { return p.serial == serial; });

    if (acked_event == std::end(inflight_configures))
    {
        BOOST_THROW_EXCEPTION(std::runtime_error(
            "Could not find acked configure with serial " + std::to_string(serial)));
    }

    auto const acked_configure_size = acked_event->second;

    inflight_configures.erase(std::begin(inflight_configures), ++acked_event);
    
    if (!inflight_configures.empty())
    {
        // If a stale configure has been acked, we do no more
        return;
    }
```

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]

**Review by AlanGriffiths — APPROVED** [review]
WFM

**Pull request merged**